### PR TITLE
Write about the chain cover a little.

### PR DESCRIPTION
**Pull request by DMRobertson**
Was helpful for me.

### CODE REVIEW

File: docs/auth_chain_difference_algorithm.md [hunk 1]
```diff
@@ -34,13 +34,38 @@ the process of indexing it).
...
+> reachability queries](https://doi.org/10.1109/ICDE.2008.4497498),
+> in: 2008 IEEE 24th International Conference on Data Engineering, April 2008,
+> pp. 893–902. (PDF available via [Google Scholar](https://scholar.google.com/scholar?q=Y.%20Chen,%20Y.%20Chen,%20An%20efficient%20algorithm%20for%20answering%20graph%20reachability%20queries,%20in:%202008%20IEEE%2024th%20International%20Conference%20on%20Data%20Engineering,%20April%202008,%20pp.%20893902.).)
+
+for a more modern take.
+
+In practical terms, the chain cover assigns every event a
+*chain ID* and *sequence number* (e.g. `(5,3)`), and maintains a map of *links*
+between chains (e.g. `(5,3) -> (2,4)`) such that `A` is reachable by `B` (i.e. `A`
 is in the auth chain of `B`) if and only if either:
```
Comment (squahtx) [inline docs/auth_chain_difference_algorithm.md:66]:
```suggestion
In practical terms, the chain cover assigns every event a
*chain ID* and *sequence number* (e.g. `(5,3)`), and maintains a map of *links*
between events in chains (e.g. `(5,3) -> (2,4)`) such that `A` is reachable by `B`
(i.e. `A` is in the auth chain of `B`) if and only if either:
```

File: docs/auth_chain_difference_algorithm.md [hunk 2]
```diff
@@ -34,13 +34,38 @@ the process of indexing it).
...
 
-1. A and B have the same chain ID and `A`'s sequence number is less than `B`'s
+1. `A` and `B` have the same chain ID and `A`'s sequence number is less than `B`'s
    sequence number; or
 2. there is a link `L` between `B`'s chain ID and `A`'s chain ID such that
    `L.start_seq_no` <= `B.seq_no` and `A.seq_no` <= `L.end_seq_no`.
```
Comment (squahtx) [inline docs/auth_chain_difference_algorithm.md:71]: This bit's still a little unclear (but I did manage to understand it in the end).

Up until point 2, I wasn't sure whether sequence numbers were namespaced to chains and it was unclear why links included a sequence number.

Not sure how to improve it though.

File: docs/auth_chain_difference_algorithm.md [hunk 3]
```diff
@@ -34,13 +34,38 @@ the process of indexing it).
...
+like "is event `A` in the auth chain of event `B`?". We could do this with an index
+that tracks all pairs `(A, B)` such that `A` is in the auth chain of `B`. However this
+would be prohibitively large, scaling poorly as the room accumulates more state
+events.
+
+Instead, we break down the graph into *chains*. A chain is a subset of a DAG
+with the following property: for any pair of events `E` and `F` in the chain,
+the chain contains a path `E -> F` or a path `F -> E`. If we ensure that each
+persisted event belongs to exactly one chain, we can keep answer reachability
+queries by tracking of how the chains are connected to one another. Doing so
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply docs/auth_chain_difference_algorithm.md:47]: Oops. How about:

```suggestion
the chain contains a path `E -> F` or a path `F -> E`. Synapse ensures that each
persisted event belongs to exactly one chain, and tracks how the chains are
connected to one another. This allows us to efficiently answer reachability
queries. Doing so
```

[the remark this answers is not in the document]
Comment (squahtx) [reply docs/auth_chain_difference_algorithm.md]: Looks reasonable! Let's re-line wrap the following lines as well.

File: docs/auth_chain_difference_algorithm.md (hunk 2 above)
Comment (squahtx) [reply docs/auth_chain_difference_algorithm.md]: Makes sense! Turns out I was _really_ confused about sequence numbers. Somehow I thought that 2. implied they were global.

**Review by squahtx — APPROVED** [review]
LGTM, thanks for documenting all this!

**Comment by DMRobertson** [comment]
Sorry @squahtx, I've made one last tweak in 2b1216ac4. Could you take a look again? Last one, I promise!

File: docs/auth_chain_difference_algorithm.md [hunk 4]
```diff
@@ -34,13 +34,45 @@ the process of indexing it).
...
+for the auth chain in a room, allowing us to efficiently make reachability queries
+like "is event `A` in the auth chain of event `B`?". We could do this with an index
+that tracks all pairs `(A, B)` such that `A` is in the auth chain of `B`. However, this
+would be prohibitively large, scaling poorly as the room accumulates more state
+events.
+
+Instead, we break down the graph into *chains*. A chain is a subset of a DAG
+with the following property: for any pair of events `E` and `F` in the chain,
+the chain contains a path `E -> F` or a path `F -> E`. This forces a chain to be
+linear (without forks) e.g. `E -> F -> G -> ... -> H`. Each event in the chain
```
Comment (squahtx) [inline docs/auth_chain_difference_algorithm.md:46]:
```suggestion
linear (without forks), e.g. `E -> F -> G -> ... -> H`. Each event in the chain
```

**Review by squahtx — COMMENTED** [review]
lgtm!

**Comment by DMRobertson** [comment]
Thanks so much!

**Pull request merged**